ownership/views.py

```diff
@@ -1,194 +1,279 @@
+        message = "This project has already been processed: {}.".format(status)
+        self.set_error_details(400, message)
+
+    def process_accept(self):
+        delivery = self.delivery
+        request = self.request
+        if delivery.is_complete():
+            self._set_already_complete_error()
+            return
         try:
-            details = DeliveryDetails.from_transfer_id(transfer_id, request.user)
-            delivery = details.get_delivery()
-            # update the status
-            if delivery.is_complete():
-                return render_already_complete(request, delivery)
-            return func(request, delivery)
-        except ObjectDoesNotExist:
-            return general_error(request, msg=TRANSFER_ID_NOT_FOUND, status=404)
-        except DataServiceError as err:
-            return general_error(request, msg=(err), status=500)
-    else:
-        return general_error(request, msg=MISSING_TRANSFER_ID_MSG, status=400)
-
-
-def render_already_complete(request, delivery):
+            delivery_util = self.delivery_type.make_delivery_util(delivery, request.user)
+            delivery_util.accept_project_transfer()
+            delivery_util.share_with_additional_users()
+            warning_message = delivery_util.get_warning_message()
+            message = self.delivery_type.make_processed_message(delivery, request.user, 'accepted', warning_message=warning_message)
+            self.warning_message = warning_message
+            message.send()
+            delivery.mark_accepted(request.user.get_username(), message.email_text)
+        except DataServiceError as e:
+            self.set_error_details(500, 'Unable to transfer ownership: {}'.format(e.message))
+        except Exception as e:
+            self.set_error_details(500, str(e))
+
+    def handle_post(self):
+        request = self.request
+        if 'decline' in request.POST:
+            # Redirect to decline page
+            self.set_redirect('ownership-decline')
+        else:
+            # Cannot redirect to a POST, so we must process the acceptance here
+            self.set_redirect('ownership-accepted')
+            self.process_accept()  # May override response type with an error
+
+
+class DeclineView(DeliveryViewBase):
     """
-    User is trying to access a delivery that has already been declined or accepted.
+    Handles GET to display form to prompt for reason
+    When POSTed, process the decline action
     """
-    status = State.DELIVERY_CHOICES[delivery.state][1]
-    message = "This project has already been processed: {}.".format(status)
-    return general_error(request, msg=message, status=400)
+    http_method_names = ['get', 'post']
+    template_name = 'ownership/decline_reason.html'
+
+    def process_decline(self, reason):
+        delivery = self.delivery
+        request = self.request
+        if delivery.is_complete():
+            self._set_already_complete_error()
```

Comment:
There is no implementation of `_set_already_complete_error` in this object.
This causes a 500 if you attempt to re-decline an already declined delivery.